benchmarks/polys.py

```diff
@@ -1,17 +1,115 @@
+        self.values['Polyprem_case1'] = fp.prem(gp)
+
+    def time_PolyElement_prem_case1(self):
+            self.R = ZZ[self.x, self.y1, self.y2, self.y3, self.y4, self.y5, self.y6, self.y7, self.y8, self.y9, self.y10]
+            fpe, gpe = self.R.from_sympy(self.f), self.R.from_sympy(self.g)
+            self.values['PolyElement_prem_case1'] = fpe.prem(gpe)
+
+    def generate_poly_case2(self):
+        # Sparse GCD and inputs where degree are proportional to the number of variables.
+        n = self.size
+        D = 1 + self.x ** (n + 1) + sum([self.y[i] ** (n + 1) for i in range(n)])
+        self.f = D * (-2 + self.x ** (n + 1) + sum([self.y[i] ** (n + 1) for i in range(n)]))
+        self.g = D * (2 + self.x ** (n + 1) + sum([self.y[i] ** (n + 1) for i in range(n)]))
+
+    def time_prem_case2(self):
+        self.values['prem_case2'] = prem(self.f, self.g, self.x)
+
+    def time_Polyprem_case2(self):
+        n = self.size
+        fp, gp = Poly(self.f, self.x, *(self.y)[:n]), Poly(self.g, self.x, *(self.y)[:n])
+        self.values['Polyprem_case2'] = fp.prem(gp)
+
+    def time_PolyElement_prem_case2(self):
+        self.R = ZZ[self.x, self.y1, self.y2, self.y3, self.y4, self.y5, self.y6, self.y7, self.y8, self.y9, self.y10]
+        fpe, gpe = self.R.from_sympy(self.f), self.R.from_sympy(self.g)
+        self.values['PolyElement_prem_case2'] = fpe.prem(gpe)
+
+    def generate_poly_case3(self):
+        # Quadratic non-monic GCD, F and G have other quadratic factors.
+        n = self.size
+        D = 1 + self.x**2 * self.y[0]**2 + sum([self.y[i]**2 for i in range(1, n)])
+        self.f = D * (-1 + self.x**2 - self.y[0]**2 + sum([self.y[i]**2 for i in range(1, n)]))
+        self.g = D * (2 + self.x * self.y[0] + sum(self.y[1:n]))**2
+
+    def time_prem_case3(self):
+        self.values['prem_case3'] = prem(self.f, self.g, self.x)
+
+    def time_Polyprem_case3(self):
+        n = self.size
+        fp, gp = Poly(self.f, self.x, *(self.y)[:n]), Poly(self.g, self.x, *(self.y)[:n])
+        self.values['Polyprem_case3'] = fp.prem(gp)
+
+    def time_PolyElement_prem_case3(self):
+        self.R = ZZ[self.x, self.y1, self.y2, self.y3, self.y4, self.y5, self.y6, self.y7, self.y8, self.y9, self.y10]
+        fpe, gpe = self.R.from_sympy(self.f), self.R.from_sympy(self.g)
+        self.values['PolyElement_prem_case3'] = fpe.prem(gpe)
+
+    def generate_poly_case4(self):
+        # Sparse non-monic quadratic inputs with linear GCDs.
+        n = self.size
+        D = -1 + self.x * prod(self.y[:n])
+        self.f = D * (3 + self.x * prod(self.y[:n]))
+        self.g = D * (-3 + self.x * prod(self.y[:n]))
+
+    def time_prem_case4(self):
+        self.values['prem_case4'] = prem(self.f, self.g, self.x)
+
+    def time_Polyprem_case4(self):
+        n = self.size
+        fp, gp = Poly(self.f, self.x, *(self.y)[:n]), Poly(self.g, self.x, *(self.y)[:n])
+        self.values['Polyprem_case4'] = fp.prem(gp)
```

Comment:
It is important not to put anything in the `time_` method except for the actual thing that we want to time which in this case is the call to `fp.prem`. Anything else will affect the timings recorded by asv. That is why there is a separate `setup` method so that we can setup the objects needed but not include the time for setup in the measured timings.

The `time_prem_case4` method is good because it only times the actual call to `prem`.

The `time_Polyprem_case4` method is not because it first calls `Poly` and then calls `prem`. The creation of the two `Poly` objects should happen in the `setup` method and not in the `time_` method.